src/async_client.rs

```diff
@@ -565,9 +562,8 @@ impl AsyncClient {
         let mut response = self.send(request).await?;
 
         for (room_id, room) in &mut response.rooms.join {
-            let mut client = self.base_client.write().await;
-
             let _matrix_room = {
+                let mut client = self.base_client.write().await;
                 for event in &room.state.events {
                     if let EventResult::Ok(e) = event {
                         client.receive_joined_state_event(&room_id, &e).await;
```

Comment:
Moving the lock of `BaseClient` into the smallest scope possible is okay right. When I made this change that's when the example started working.